internal/validators/validate_test.go

```diff
@@ -0,0 +1,162 @@
+				"requiredArrayField": "Should have at least 1 element",
+				"enumField":          `Unexpected value "invalid". Expected one of the following values: foo, bar`,
+				"publicKeyField":     "Invalid public key provided",
+			},
+		},
+		{
+			stc: &testStruct{
+				RequiredField:      "foo",
+				RequiredArrayField: []string{"bar", ""},
+				EnumField:          "bar",
+				PublicKeyField:     keypair.MustRandom().Address(),
+				UnknownTagField:    1,
+			},
+			expectedFieldErrors: map[string]interface{}{
+				"requiredArrayField[1]": "This field cannot be empty",
+			},
+		},
+		{
+			stc: &testStruct{
+				RequiredField:      "foo",
+				RequiredArrayField: []string{"bar"},
+				EnumField:          "bar",
+				PublicKeyField:     keypair.MustRandom().Address(),
+				UnknownTagField:    2,
+			},
+			expectedFieldErrors: map[string]interface{}{
+				"unknownTagField": "Invalid value",
+			},
+		},
+		{
+			stc: &testStruct{
+				RequiredField:      "foo",
+				RequiredArrayField: []string{"bar"},
+				EnumField:          "bar",
+				PublicKeyField:     keypair.MustRandom().Address(),
+				UnknownTagField:    1,
+				NestedField: []testStructNested{
+					{
+						NestedRequiredField: "",
+						NestedEnumField:     "invalid",
+					},
+				},
+			},
+			expectedFieldErrors: map[string]interface{}{
+				"nestedField[0].nestedRequiredField": "This field is required",
+				"nestedField[0].nestedEnumField":     `Unexpected value "invalid". Expected one of the following values: foo, bar`,
+			},
+		},
+	}
+
+	val := NewValidator()
+	for _, tc := range testCases {
+		err := val.Struct(tc.stc)
+		require.Error(t, err)
+		vErrs, ok := err.(validator.ValidationErrors)
+		require.True(t, ok)
+		fieldErrors := ParseValidationError(vErrs)
+		assert.Equal(t, tc.expectedFieldErrors, fieldErrors)
+	}
+}
+
+func TestGetFieldName(t *testing.T) {
+	type testStructNested struct {
+		Name     string             `validate:"not_empty"`
+		Children []testStructNested `validate:"dive"`
+	}
+
+	type testStruct struct {
+		PublicKey   string             `validate:"public_key"`
+		NestedField []testStructNested `validate:"required,dive"`
+	}
+
+	stc := &testStruct{
+		PublicKey: "",
+		NestedField: []testStructNested{
+			{
+				Name: "first",
+				Children: []testStructNested{
+					{
+						Name: "second",
+						Children: []testStructNested{
+							{
+								Name:     "children1",
+								Children: []testStructNested{},
+							},
+							{
+								Name: "children2",
+								Children: []testStructNested{
+									{
+										Name:     "",
+										Children: []testStructNested{},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	val := NewValidator()
+	err := val.Struct(stc)
+	require.Error(t, err)
+
+	vErrs, ok := err.(validator.ValidationErrors)
+	require.True(t, ok)
+	require.Len(t, vErrs, 2)
+
+	assert.Equal(t, "publicKey", getFieldName(vErrs[0]))
+	assert.Equal(t, "children[0].name", getFieldName(vErrs[1]))
```

Comment:
+1

Reply:
There is no particular reason, I can make it to return the entire path except for the Root struct (otherwise we'll leak an internal structure). I thought that only showing the last two would make it more clear.